editoast/src/modelsv2/prelude/list.rs

```diff
@@ -43,14 +44,17 @@ impl<M: Model> FilterSetting<M> {
 pub struct SortSetting<M: Model>(pub(in crate::modelsv2) DieselSort<M::Table>);
 type DieselSort<Table> = Box<dyn diesel::BoxableExpression<Table, Pg, SqlType = NotSelectable>>;
 
-/// A builder for model listing settings
+/// A builder struct to accumulate settings that define a
+/// selection of [Model](super::Model) objects
 ///
-/// See [List]
+/// Used by the [List] and [Count] traits to determine which rows
+/// to consider.
 pub struct SelectionSettings<M: Model> {
-    pub(in crate::modelsv2) filters: Vec<FilterSetting<M>>,
-    pub(in crate::modelsv2) sorts: Vec<SortSetting<M>>,
+    pub(in crate::modelsv2) filters: Vec<Arc<dyn Fn() -> FilterSetting<M> + Send + Sync>>,
```

Comment:
If I understand, we can't have `FilterSetting` be clonabled because of `diesel::BoaxableExpression` which is not.

One way of bypassing that is not to store the object itself, but a factory of the object (a function that can be called to create the object each time it's called).

Another way would be to wrap `FilterSetting` in an `Arc`.

But I do not understand why you do both at the same time (wrapping the factory in an `Arc`). What did I miss?

Also, the function does make the general API less easy to use, this is more straightforward:

```rust
SelectionSettings::new().filter(Document::CONTENT_TYPE.eq(json))
```
than this

```rust
SelectionSettings::new().filter(move || Document::CONTENT_TYPE.eq(json.to_string()))
```

Reply:
The problem is that in the implementation of the traits by the macro, we do:

```rust
for filter_fun in settings.filters {
    let crate::modelsv2::prelude::FilterSetting(filter) = (*filter_fun)();
    query = query.filter(filter);
}
```

in which `query.filter(filter)` requires ownership of `filter` (if we could pass a reference we'd be taking a `&SelectionSettings<Model>` in parameter of the function). So even if `Box<dyn BoxableExpression>` was `Send + Sync`, we'd still need ownership of the pointee of the `Arc<FilterSetting>` to build the expression, thus potentially breaking other copies of the `Arc`. (Or maybe I'm misunderstanding something about `Arc`, which tbh is quite possible 😬)

Even though I agree with you about the API, we'll mostly use `.order_by` which doesn't require any capture so it's literally just an extra `|| ` to pay. I don't believe we'll use `.filter` in many places.